cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/JarCacheTracker.java

```diff
@@ -84,96 +84,90 @@ public void registerLaunch(File cacheDir, ProgramType programType) {
 
     // synchronized here instead of at registerLaunch so that we only per program type.
     private synchronized void waitForCacheFiles(File cacheDir) {
+      Path cachePath = cacheDir.getAbsoluteFile().toPath();
+      Path containerDonePath = cachePath.resolve("container.jar.done");
+      Path appmasterDonePath = cachePath.resolve("appMaster.jar.done");
+
       // check this in case we called this method before the files were available, were blocked on another thread,
       // and are now here
       long threshold = firstLaunchTime.get() + TimeUnit.MILLISECONDS.convert(90, TimeUnit.SECONDS);
-      if (System.currentTimeMillis() > threshold || cacheFilesExist(cacheDir)) {
+      if (System.currentTimeMillis() > threshold ||
+        (Files.exists(containerDonePath) && Files.exists(appmasterDonePath))) {
         return;
       }
 
       // otherwise, wait for at most threshold seconds for the cache files to exist
-      WatchService watcher;
-      WatchKey key;
-      try {
-        watcher = FileSystems.getDefault().newWatchService();
-        key = cacheDir.toPath().register(watcher, StandardWatchEventKinds.ENTRY_CREATE);
-      } catch (Exception e) {
-        LOG.error("Error watching {} for twill cache files to exist. Continuing on.", cacheDir, e);
-        return;
-      }
-
-      // if the cache files were written while we were registering the watcher
-      Path containerDonePath = FileSystems.getDefault().getPath(cacheDir.getAbsolutePath(), "container.jar.done");
-      Path appmasterDonePath = FileSystems.getDefault().getPath(cacheDir.getAbsolutePath(), "appMaster.jar.done");
-      boolean containerJarExists = Files.exists(containerDonePath);
-      boolean appmasterJarExists = Files.exists(appmasterDonePath);
-      if (containerJarExists && appmasterJarExists) {
-        LOG.debug("files exist, cancelling watch key.");
-        key.cancel();
-        return;
-      }
-
-      LOG.debug("Waiting for twill cache files at {}", cacheDir);
-      while (!containerJarExists || !appmasterJarExists) {
-
-        // wait at most threshold millis for some event to occur
-        long maxWaitTime = threshold - System.currentTimeMillis();
-        if (maxWaitTime < 0) {
-          break;
-        }
+      try (WatchService watcher = FileSystems.getDefault().newWatchService()) {
+        WatchKey key;
         try {
-          key = watcher.poll(maxWaitTime, TimeUnit.MILLISECONDS);
+          key = cachePath.register(watcher, StandardWatchEventKinds.ENTRY_CREATE);
```

Comment:
this part is really confusing due to how diffs show indentation changes, but the only difference in this block is that WatcherService creation is in a try(), and Paths are being compared instead of Strings.